### PR TITLE
Handle first server SETTINGS frame async

### LINKED ISSUE #311: HTTP/2 waiting for server settings increases latency
In HTTP2 it is optional for the client to wait for the server settings frame before making requests. However, Mint waits for the server settings frame and this increases the amount of time before Mint is able to make a request. For example if it takes 80ms for a packet to be sent from the server to the client then this increases the latency for the first request by 80ms. If you look at browser implementations like Chrome, Firefox or Safari none of these browsers will wait for the server settings frame before making a request.

### PR DESCRIPTION
**Pull request by whatyouhide**
Closes #311.

### CODE REVIEW

File: lib/mint/http2.ex [hunk 1]
```diff
@@ -429,6 +429,7 @@ defmodule Mint.HTTP2 do
   def open?(%Mint.HTTP2{state: state} = _conn, type \\ :read_write)
       when type in [:read, :write, :read_write] do
     case state do
+      :handshaking -> true
```
Comment (whatyouhide) [inline lib/mint/http2.ex:432]: New state for the connection.

File: lib/mint/http2.ex [hunk 2]
```diff
@@ -1438,18 +1399,42 @@ defmodule Mint.HTTP2 do
...
+        conn.state == :handshaking and type != :settings ->
+          debug_data = "received invalid frame #{type} during handshake"
+          send_connection_error!(conn, :protocol_error, debug_data)
+
+        conn.state == :handshaking ->
+          %__MODULE__{conn | state: :open}
+
+        true ->
+          conn
+      end
```
Comment (whatyouhide) [inline lib/mint/http2.ex:1432]: Important code is here.

File: test/mint/http2/conn_test.exs [hunk 3]
```diff
@@ -19,7 +19,15 @@ defmodule Mint.HTTP2Test do
...
-    quote do: refute_receive({:ssl, _socket, _data})
+    quote do
+      receive do
+        {:ssl, _socket, data} ->
+          result = Mint.HTTP2.Frame.decode_next(data)
+          flunk("Expected no frames, but got data that decodes to: #{inspect(result)}")
+      after
+        100 -> :ok
+      end
+    end
```
Comment (whatyouhide) [inline test/mint/http2/conn_test.exs:30]: Quality of life improvement

File: test/support/mint/http2/test_server.ex [hunk 4]
```diff
@@ -16,28 +17,61 @@ defmodule Mint.HTTP2.TestServer do
...
-    # SETTINGS here.
     conn =
-      if options[:mode] == :passive do
-        assert {:ok, %HTTP2{} = conn, []} = HTTP2.recv(conn, 0, 100)
-        conn
-      else
-        assert_receive message, 100
-        assert {:ok, %HTTP2{} = conn, []} = HTTP2.stream(conn, message)
-        conn
+      case Keyword.get(options, :first_server_frame, :settings) do
```
Comment (whatyouhide) [inline test/support/mint/http2/test_server.ex:38]: Had to make some hacky changes to the test server in order to test this.

File: lib/mint/http2.ex [hunk 5]
```diff
@@ -1438,18 +1399,42 @@ defmodule Mint.HTTP2 do
...
 
-  defp assert_valid_frame(_conn, unknown()) do
+  defp assert_valid_frame(conn, unknown()) do
     # Unknown frames MUST be ignored:
     # https://datatracker.ietf.org/doc/html/rfc7540#section-4.1
-    :ok
+    conn
   end
 
   defp assert_valid_frame(conn, frame) do
```
Comment (ericmj) [inline lib/mint/http2.ex:1408]: Should this function be renamed since it also updates connection state now?

**Review by ericmj — APPROVED** [review]

File: lib/mint/http2.ex (hunk 5 above)
Comment (whatyouhide) [reply lib/mint/http2.ex]: Good call, done! ✅

**Pull request merged**